### PR TITLE
Update of TPC Splines: non-uniform grid for TPC residuals

### PR DESCRIPTION
**Pull request by sgorbuno**
The PR contains several commits:

- non-uniform grid that corresponds to the track residual voxels

-  limits for SC correction values per TPC row
-  multithreaded reading of the residual tree
- fix propagation of the track residual data to the TPC row edges
-  fix initialization of the track residuals in the test macro

### CODE REVIEW

**Comment by vkucera** [comment]
Seems like fullCI is having some unrelated problems.

**Review by wiechula — CHANGES_REQUESTED** [review]
@sgorbuno , please
```c++
#include <algorithm>
```
in `TPCFastSpaceChargeCorrection.h`,  rebase and force push.

**Comment by wiechula** [comment]
@sgorbuno , unfortunately this did not work. I guess code parts that are not supposed to run on GPU need protection. I think default header files will not work, I overlooked that. If min / max etc. is supposed to run also on GPU you could use https://github.com/AliceO2Group/AliceO2/blob/dev/GPU/Common/GPUCommonMath.h#L331. Please check what would be appropriate.

**Comment by sgorbuno** [comment]
Hi @wiechula , 

I'm not allowed to see the build log when clicking on "Details":  "Unauthorized to log in. A missing role is required to access this application. Please contact the application owner to get access."  Can it be fixed?

**Comment by shahor02** [comment]
@sgorbuno this is strange, perhaps @ktf can tell why you cannot see the CI details. But in any case, the failure on macOS-arm is unrelated to this PR.

File: GPU/TPCFastTransformation/macro/TPCFastTransformInit.C [hunk 1]
```diff
@@ -21,19 +21,24 @@
 /// root -l TPCFastTransformInit.C'("debugVoxRes.root")'
 ///
 
+#include "Algorithm/RangeTokenizer.h"
+
```
Comment (wiechula) [inline GPU/TPCFastTransformation/macro/TPCFastTransformInit.C]:
```suggestion
```
Already below

File: GPU/TPCFastTransformation/macro/TPCFastTransformInit.C [hunk 2]
```diff
@@ -75,30 +82,107 @@ void TPCFastTransformInit(const char* fileName = "debugVoxRes.root",
...
+      std::cout << "z2x: " << z2x << std::endl;
+    }
+    std::cout << std::endl;
+
+    int nZ2Xbins = trackResiduals.getNZ2XBins();
+    std::cout << " TrackResiduals z2x bins: " << nZ2Xbins << std::endl;
+    for (int i = 0; i < nZ2Xbins; i++) {
+      std::cout << "getZ2X(bin) : " << trackResiduals.getZ2X(i) << std::endl;
+    }
+  }
```
Comment (wiechula) [inline GPU/TPCFastTransformation/macro/TPCFastTransformInit.C]: Is this for debugging? Perhaps add a comment.

File: GPU/TPCFastTransformation/macro/TPCFastTransformInit.C [hunk 3]
```diff
@@ -75,30 +82,107 @@ void TPCFastTransformInit(const char* fileName = "debugVoxRes.root",
...
+          std::cout << "roc " << iRoc << " row " << iRow
+                    << " gridV0 " << info.gridV0 << " gridCorrU0 " << info.gridCorrU0 << " gridCorrV0 " << info.gridCorrV0
+                    << " scaleCorrUtoGrid " << info.scaleCorrUtoGrid << " scaleCorrVtoGrid " << info.scaleCorrVtoGrid
+                    << " gridU0 " << info.gridU0 << " scaleUtoGrid " << info.scaleUtoGrid << " scaleVtoGrid " << info.scaleVtoGrid
+                    << std::endl;
+        }
+      }
+    }
+  }
+
```
Comment (wiechula) [inline GPU/TPCFastTransformation/macro/TPCFastTransformInit.C:184]: Why are the data written to file and read back?
Is this only for debugging? If so, please add a comment.

File: GPU/TPCFastTransformation/macro/TPCFastTransformInit.C [hunk 4]
```diff
@@ -125,12 +209,21 @@ void TPCFastTransformInit(const char* fileName = "debugVoxRes.root",
...
 
+  // duplicate of debugVox
+  debugFile->cd();
+  TNtuple* debugCorrVox =
+    new TNtuple("corrvox", "corrvox", "iRoc:iRow:n:x:y:z:vx:vy:vz:cx:cy:cz");
+
+  debugCorrVox->SetMarkerStyle(8);
+  debugCorrVox->SetMarkerSize(0.8);
+  debugCorrVox->SetMarkerColor(kMagenta);
+
```
Comment (wiechula) [inline GPU/TPCFastTransformation/macro/TPCFastTransformInit.C:229]: Why is this added? It indeed seems to be an exact copy of `debugVox` filled with the same entries below.

**Review by wiechula — COMMENTED** [review]
Thanks @sgorbuno !
Just a few minor comments.

File: GPU/TPCFastTransformation/macro/TPCFastTransformInit.C (hunk 4 above)
Comment (sgorbuno) [reply GPU/TPCFastTransformation/macro/TPCFastTransformInit.C:229]: The only reason is to have the tree in two different colors and marker sizes.  One can then draw vx,vy, and vz in red color and cx,cy, and cz in magenta without much typing in the command line. One can put these values in different trees, but sometimes I want to see cx-vx, cy-vy etc.

**Comment by sgorbuno** [comment]
@wiechula, I mark the PR draft again. Alexander Schmah sees some problems; It makes sense to fix them first.

File: GPU/TPCFastTransformation/macro/TPCFastTransformInit.C (hunk 1 above)
Comment (sgorbuno) [reply GPU/TPCFastTransformation/macro/TPCFastTransformInit.C]: @wiechula , 

the macro does not work for me without this header.  

the interpreter does not understand this line:
 auto y2xBins = o2::RangeTokenizer::tokenize<float>(userInfo->FindObject("y2xBinning")->GetTitle());

**Comment by sgorbuno** [comment]
@wiechula , 

I fixed Alexander's problem; one can merge the PR.  
Can you please remove the "draft" mark? I didn't manage to remove it.

> @wiechula, I mark the PR draft again. Alexander Schmah sees some problems; It makes sense to fix them first.

File: GPU/TPCFastTransformation/macro/TPCFastTransformInit.C (hunk 1 above)
Comment (wiechula) [reply GPU/TPCFastTransformation/macro/TPCFastTransformInit.C]: @sgorbuno , yes, but now it is included twice. I think the macro should be used compiled, then the `#if ...` should include it. I never had problems with it.

File: GPU/TPCFastTransformation/macro/TPCFastTransformInit.C (hunk 3 above)
Comment (wiechula) [reply GPU/TPCFastTransformation/macro/TPCFastTransformInit.C:184]: @sgorbuno , can you comment on this?

**Comment by wiechula** [comment]
@sgorbuno , thanks a lot for the update! We'll have a look first if everything works as expected, then we can proceed.
